### ISSUE TITLE
Keyboard shortcuts are captured when out of focus

### ISSUE THREAD

**Issue by karthanistyr** [skyjake/lagrange#10]
**Issue:**
When Lagrange is out of focus, I press a key combination that would be recognised by Lagrange, and Lagrange acts on it.

In this video (https://video.barbed.fr/videos/watch/fe80dc58-8807-4d3f-9850-74a19b1baf11), the top right terminal is focused (mouse pointer is over it). I press Ctrl+D to close the terminal, which has the side effect of triggering the Bookmark function in Lagrange, which was unfocused at the time of pressing.

**Expected behaviour:**
I would not expect Lagrange to trigger when a key combination is pressed when the Lagrange window is out of focus.

**Platform:**
Arch Linux x64
window manager: dwm (https://dwm.suckless.org/)

**Lagrange version:**
Self-compiled, 0.6 (commit: bcca576f0463df67378ece8712dc0941c67f0c9e)

**Comment by skyjake** [comment]
How does dwm handle input focus? Is it based on mouse cursor position? (Haven't used it myself, sorry.)

Lagrange relies entirely on SDL for input events — there is no custom behavior here. When I understand exactly what is happening here I can introduce some options to improve behavior in this case, although I expect them to be disabled by default.

**Comment by karthanistyr** [comment]
I accept the tiling wms aren't exactly the norm, however it's the first program that I saw behaving like this on my machine.

dwm does focus the window under the mouse cursor or alternatively cycles focus between visible windows using Super+J and Super+K

Actually I tried using a different combination (Ctrl+= to increase Zoom and Ctrl+0 to reset it) and couldn't replicate Lagrange acting on them while unfocused (in this new test, the number of layout of windows isn't changed, contrary to the video above). So I guess I'm looking at some form of race condition, as the Lagrange window was brought into focus when the terminal window was closed in the video test.

**Comment by skyjake** [comment]
Hmm, perhaps the best workaround would then be to allow changing Lagrange's keybindings? That way you could use keys that don't overlap with dwm's bindings.

**Comment by karthanistyr** [comment]
Ctrl+D isn't a dwm binding, it sends EOF on stdin, which zsh (and many others) interprets as an exit command on an empty line.

> perhaps the best workaround would then be to allow changing Lagrange's keybindings

It's definitely a workaround; please bear in mind that this ticket is not a showstopper, more like a quirk, so there isn't an urgency to fix any of it. I do wonder, though, wether this is a behaviour seen on other tiling wms or just dwm. I'll try some others and get back to you.

**Comment by karthanistyr** [comment]
I replicated it on openbox (non tiling wm)

- Ctrl+D: https://video.barbed.fr/videos/watch/6430e4ce-0d2f-451d-b369-ff49c179dc5b
- Ctrl+W: a bit more worrying, as it potentially exits lagrange https://video.barbed.fr/videos/watch/3348b1ae-cf5d-4f41-a50b-18fd1f0d5931

These behaviours are triggered when the lagrange window comes into focus after the intended target window of the key binding closes

**Comment by skyjake** [comment]
> please bear in mind that this ticket is not a showstopper, more like a quirk, so there isn't an urgency to fix any of it.

I agree, but seeing as I had overlooked the expose event I wanted to look into other window events that may require additional handling.

I checked the information in the SDL keydown events, and the Ctrl+D received by Lagrange following xterm closing is identical to a normal keydown event. Could be some glitch in the window manager event handling where the event is considered not eaten by the closed window, and thus gets passed on to the next one.

Applied a workaround although it is a bit of a kludge...
